SnowplowTracker/SnowplowTracker/Emitter.cs

```diff
@@ -0,0 +1,528 @@
+		                int sendLimit = 500, int bufferSize = 25, long byteLimitGet = 52000, long byteLimitPost = 52000) {
+			Utils.CheckArgument (!String.IsNullOrEmpty (endpoint), "Endpoint cannot be null or empty.");
+			this.endpoint = endpoint;
+			this.collectorUri = MakeCollectorUri(endpoint, protocol, method);
+			this.httpProtocol = protocol;
+			this.httpMethod = method;
+			this.sendLimit = sendLimit;
+			this.bufferSize = bufferSize;
+			this.byteLimitGet = byteLimitGet;
+			this.byteLimitPost = byteLimitPost;
+			this.eventStore = new EventStore ();
+		}
+
+		/// <summary>
+		/// Adds an event payload to the database.
+		/// </summary>
+		/// <param name="payload">Payload.</param>
+		public void Add(TrackerPayload payload) {
+			payloadQueue.Enqueue (payload);
+		}
+
+		/// <summary>
+		/// Starts the Emitter and the Event Consumer.
+		/// - EmitLoop will send all events in the database and then
+		///   wait for the event consumer to signal that there is work.
+		/// - The event consumer waits for its queue to be signalled and
+		///   will then add an event to the database; after which it 
+		///   signals the emitloop of work to be done.
+		/// </summary>
+		public void Start() {
+			if (!sending) {
+				// Start the emitter sending loop.
+				emitThread = new Thread(new ThreadStart(EmitLoop));
+				sending = true;
+				emitThread.Start ();
+
+				// Start the event consumer.
+				payloadConsumer = new Thread(new ThreadStart(EventConsumer));
+				consuming = true;
+				payloadConsumer.Start();
+			}
+		}
+		
+		/// <summary>
+		/// Stops the Emitter and the Event Consumer.
+		/// </summary>
+		public void Stop() {
+			if (sending) {
+				// Stop the emitter sending loop.
+				sending = false;
+				lock (emitLock) {
+					Monitor.Pulse(emitLock);
+				}
+
+				// Stops the event consumer.
+				consuming = false;
+				payloadQueue.Enqueue(null);
+			}
+		}
+
+		// --- Consumers & Loopers
+
+		/// <summary>
+		/// Consumes events from the event queue and adds them to the database.
+		/// </summary>
+		private void EventConsumer() {
+			Log.Debug ("Emitter: Event consumer starting up");
+
+			while (consuming) {
+				TrackerPayload payload = payloadQueue.Dequeue ();
+
+				// If the consumer was shutdown while waiting
+				if (!consuming) {
+					Log.Debug ("Emitter: Event consumer shutting down...");
+					break;
+				}
+
+				// Signal emit loop
+				lock (emitLock) {
+					this.eventStore.AddEvent(payload);
+					Monitor.Pulse(emitLock);
+				}
+			}
+		}
+	
+		/// <summary>
+		/// Performs and orchestrates all sending of events from the database
+		/// and subsequent removal of events from the database.
+		/// </summary>
+		private void EmitLoop() {
+			Log.Debug ("Emitter: EmitLoop starting up");
+
+			while (sending) {
+				// Wait for something to be sent!
+				lock (emitLock) {
+					if (eventStore.GetEventCount() == 0) {
+						Monitor.Wait(emitLock);
+					}
+				}
```

Comment:
This looks correct to me!

One thing to be aware of when using condition variables is [spurious wakeup](https://en.wikipedia.org/wiki/Spurious_wakeup). This means that it is theoretically possible for the sleeping thread waiting on the monitor to be awoken even without a pulse. The normal way to deal with this is by checking the think you were waiting for in a loop (i.e. using `while` rather than `if`).

In this case I don't think it makes a difference - if the thread is spuriously awoken then it will get a list of events, find that the list is empty, and go back to sleep.

http://stackoverflow.com/questions/1461913/does-c-sharp-monitor-wait-suffer-from-spurious-wakeups